### PR TITLE
An implementation of CGS

### PR DESCRIPTION
**Pull request by Xx-fractale-xX**
Classic version of CGS with right preconditioning.
cgs

### CODE REVIEW

File: src/cgs.jl [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+it does not involve adjoint matrix-vector multiplications, and the expected convergence
+rate is about twice that of the BCG algorithm.
+
+The Conjugate Gradient Squared algorithm works quite well in many cases.
+However, one difficulty is that, since the polynomials are squared, rounding errors
+tend to be more damaging than in the standard BCG algorithm. In particular, very
+high variations of the residual vectors often cause the residual norms computed 
+to become inaccurate. 
+
+TFQMR and BICGSTAB were developed to remedy this difficulty.
```
Comment (dpo) [inline src/cgs.jl]: Is this description copy-pasted from a paper or a book?

File: src/cgs.jl [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+  p = copy(r₀) # p₀
+
+  # Stopping criterion.
+  solved = rNorm ≤ ε
+  tired = iter ≥ itmax
+  status = "unknown"
+
+  while !(solved || tired)
+
+    # yₘ = M⁻¹pₘ
```
Comment (dpo) [inline src/cgs.jl]: Would you mind placing the comments to the right of the statements to improve readability?

File: test/test_cgs.jl [hunk 3]
```diff
@@ -0,0 +1,61 @@
+include("get_div_grad.jl")
```
Comment (dpo) [inline test/test_cgs.jl]: I think all these examples are symmetric, right? We should test CGS on unsymmetric systems.

File: src/cgs.jl (hunk 1 above)
Comment (Xx-fractale-xX) [reply src/cgs.jl]: Yes, I took it in ```Iterative methods for sparse linear systems```.

Comment (dpo) [reply src/cgs.jl]: Then there should be a sentence to cite the source. For example, you could start the docstring with `From "Iterative Methods for Sparse Linear Systems (C. T. Kelley):" ...`

File: src/cgs.jl (hunk 2 above)
Comment (Xx-fractale-xX) [reply src/cgs.jl]: Ok, I do the modifications.

File: test/test_cgs.jl (hunk 3 above)
Comment (Xx-fractale-xX) [reply test/test_cgs.jl]: Yes, I took the same tests that I used for ```DQGMRES``` and ```DIOM```. I will find two new tests, a not symmetric definite positive matrix and an other one not symmetric and indefinite.

File: src/cgs.jl (hunk 1 above)
Comment (Xx-fractale-xX) [reply src/cgs.jl]: Ok, I modify that too. Thanks!

**Comment by dpo** [comment]
This branch needs to be rebased now.

File: src/cgs.jl [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+The Conjugate Gradient Squared algorithm works quite well in many cases.
+However, one difficulty is that, since the polynomials are squared, rounding errors
+tend to be more damaging than in the standard BCG algorithm. In particular, very
+high variations of the residual vectors often cause the residual norms computed 
+to become inaccurate.
+
+TFQMR and BICGSTAB were developed to remedy this difficulty.»
+
+This implementation allows a right preconditioning with M.
```
Comment (dpo) [inline src/cgs.jl]: ...allows a right preconditioner M.

**Comment by dpo** [comment]
Looks great, thank you!

File: src/cgs.jl (hunk 4 above)
Comment (Xx-fractale-xX) [reply src/cgs.jl]: Thanks! I also updated `DIOM` and `DQGMRES`, I made the same mistake.

**Pull request merged**